### ISSUE TITLE
Publish error: ...obj\project.assets.json' doesn't have a target for '.NETCoreApp,Version=v2.2'

### ISSUE THREAD

**Issue by robinwilson16** [aspnet/AspNetCore.Docs#14705]
These instructions seem more like a set of release notes including information regarding some aspects that will need to be amended in a project and examples of how certain features work in .NET Core 3.0 compared with 2.2 but what I think it needed is a step by step list of instructions to update the project with the additional code examples and release information on another page to make the process of upgrading projects easier (e.g. open file X, find string Y, replace with string Z), more like the instructions for migrating from 2.1 to 2.2.

The instructions for migrating from .NET Core 2.1 to 2.2 are well structured and state which files to modify to upgrade a project to 2.2.
However, similar instructions to upgrade from 2.2 to 3.0 go from an estimated read time of 5 minutes (for 2.2) to 20 minutes (for 3.0) and contain a lot of additional information which may be better placed on other pages.

Often samples of code are given and it is not clear if this is just demonstrating a feature or part of the actual instructions to update your project.

Would it be possible to provide more succinct instructions which will hopefully encourage everyone to move on from 2.2?

Thanks
Robin

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 76f39eec-9dd1-6167-56ec-47b26e0b5063
* Version Independent ID: f9f2b0f3-01eb-cf58-8767-194fe0fc0e44
* Content: [Migrate from ASP.NET Core 2.2 to 3.0](https://docs.microsoft.com/en-us/aspnet/core/migration/22-to-30?view=aspnetcore-3.0&tabs=visual-studio#feedback)
* Content Source: [aspnetcore/migration/22-to-30.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/migration/22-to-30.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-migration**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

**Comment by robinwilson16** [comment]
This is what I have so far from comparing a new .NET Core 3.0 project to an existing 2.2 one although still can't manage to do a successful migration yet:

**1. Update the csproj file**
Navigate to the project directory where the solution file is located and then into the sub-folder containing the project files. Edit the .csproj file.

Find the reference to the target framework:
```
<TargetFramework>netcoreapp2.2</TargetFramework>
```

Replace with:
```
<TargetFramework>netcoreapp3.0</TargetFramework>
```

Find the reference to the hosting model. If this is set to InProcess this section can be removed as this is now the default:
```
<PropertyGroup>
    <AspNetCoreHostingModel>InProcess</AspNetCoreHostingModel>
</PropertyGroup>
```

Find and remove this line as this package is now split up into seperate packages to avoid duplication and ensure projects remain lean and efficient:
```
<PackageReference Include="Microsoft.AspNetCore.App" />
```

Remove any item groups relating to bundleconfig.json as this is no longer used.

**2. Add new dependencies:**
Open the project and go to Tools > NuGet Package Manager > Manage NuGet Packages for Solution...

Search for and install the following:
- Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore
- Microsoft.AspNetCore.Identity.EntityFrameworkCore
- Microsoft.AspNetCore.Identity.UI
- Microsoft.EntityFrameworkCore.SqlServer
- Microsoft.EntityFrameworkCore.Tools

Also install updates from the Updates tab whilst here.

**3. Update Startup.cs**
Open Startup.cs from within Visual Studio.

Find the following in the ConfigureServices section:
```
services.AddMvc().SetCompatibilityVersion(CompatibilityVersion.Version_2_2);
```

Replace with (compatibility version is no longer required):
```
services.AddRazorPages();
```

Now find the following in the Configure section:
```
app.UseCookiePolicy();
app.UseAuthentication();
app.UseMvc();
```

Replace with:
```
app.UseRouting();
app.UseAuthentication();
app.UseAuthorization();
app.UseEndpoints(endpoints =>
{
    endpoints.MapRazorPages();
});
```

**3. Update Program.cs**
Open Program.cs from within Visual Studio.

Find the following in the Main function at the top:
```
CreateWebHostBuilder(args).Build().Run();
```

Replace with:
```
CreateHostBuilder(args).Build().Run();
```

Find the following in the function below:
```
public static IWebHostBuilder CreateWebHostBuilder(string[] args) =>
    WebHost.CreateDefaultBuilder(args)
```

Replace with:
```
Host.CreateDefaultBuilder(args)
    .ConfigureWebHostDefaults(webBuilder =>
    {
        webBuilder.ConfigureKestrel(serverOptions =>
        {
            // Set properties and call methods on options
        })
```

If present find and remove:
```
.ConfigureKestrel((context, options) =>
{
    // Set properties and call methods on options
});
```

Now project is still giving lots of errors so must have missed something else.
Please let me know if you know what I have missed.

Thanks
Robin

**Comment by robinwilson16** [comment]
With the instructions above Build works and I can even run the project in IIS Express but publish fails with the following error:
```
...obj\project.assets.json' doesn't have a target for '.NETCoreApp,Version=v2.2'
```

I then tried adding this to appsettings.json which seemed the best way to solve this but this seems to cause the publish option to be stuck in a sort of loop as it never completes:
```
"targets": {
    ".NETCoreApp,Version=v3.0": {
    }
}
```

**Comment by Rick-Anderson** [comment]
@vijayrkn  after migration to 3.0 @robinwilson16  reports the following error:

```
...obj\project.assets.json' doesn't have a target for '.NETCoreApp,Version=v2.2'
```

**Comment by robinwilson16** [comment]
Ok got it working now.
Seems you need to delete the bin and obj folders and then run dotnet build and now it works and publishes.
Seems to create a lot more files but seems this is normal (comparing to a new blank .NET Core Web Application) so think this is finally solved.

**Comment by vijayrkn** [comment]
If you the app was targeting 2.2 before, then the publish profile might still have the TFM set as 2.2 

You can update this from the UI or go to Properties\PublishProfiles\<profile name>.pubxml & update it manually.

**Comment by robinwilson16** [comment]
> @vijayrkn after migration to 3.0 @robinwilson16 reports the following error:
> 
> ```
> ...obj\project.assets.json' doesn't have a target for '.NETCoreApp,Version=v2.2'
> ```

@vijayrkn @Rick-Anderson I have found this error goes away if you delete the bin and obj folders and then run `dotnet build` from the commandline then attempt to publish again from within VS. Not sure if this is a bug or something we are supposed to have to do? All seems ok now and publish is successful but have not yet uploaded to my Linux host. Will try this tomorrow.

**Comment by Rick-Anderson** [comment]
Closing for now. This is very discoverable. If other hit on this issue, I'll add it to the migration notes unless @vijayrkn  you think I should add this now.

Before publishing updated projects, delete the bin and obj folders.

**Comment by robinwilson16** [comment]
@Rick-Anderson the main reason I opened this issue was because it is very difficult to work out what you need to do to migrate an existing 2.2 project to 3.0 as you read the 20 min article and are not really any the wiser (unless that's just me?). Is it possible to consider separating the demos and release information from the steps required to upgrade the project and adding something like I proposed above which states exactly what do to and then all the other information could be on other pages? The instructions to go from 2.1>2.2 are much clearer I feel.

**Comment by robinwilson16** [comment]
> Closing for now. This is very discoverable. If other hit on this issue, I'll add it to the migration notes unless @vijayrkn you think I should add this now.
> 
> Before publishing updated projects, delete the bin and obj folders.

Hitting this issue with another project that has suddenly stopped working and deleting the `bin` and `obj` folders is having no effect.
It was a project I had already migrated to .NET Core 3.0 and had just published but then I wanted to update the version number I had stored in appsettings.json and was up to 2.1 so changed this to 2.2 and now get the error and can no longer publish further changes. Could it be this 2.2.0 has confused it and made it think I am targeting .NET Core 2.2?

Lines I amended in appsettings.json:
```
"System": {
    "Version": "2.2.0"
  }
```

Is there another way I can fix this and stop it recurring?

Thanks
Robin
